### ISSUE TITLE
ADD GLASS

### ISSUE THREAD

**Issue by kyakuno** [axinc-ai/ailia-models#1641]
https://github.com/cqylunlun/GLASS

**Comment by ooe1123** [comment]
python main.py --gpu 0 --seed 0 --test test net -b wideresnet50 -le layer2 -le layer3 --pretrain_embed_dimension 1536 --target_embed_dimension 1536 --patchsize 3 --meta_epochs 640 --eval_epochs 1 --dsc_layers 2 --dsc_hidden 1024 --pre_proj 1 --mining 1 --noise 0.015 --radius 0.75 --p 0.5 --step 20 --limit 392 dataset --distribution 0 --mean 0.5 --std 0.1 --fg 1 --rand_aug 1 --batch_size 8 --resize 288 --imagesize 288 -d carpet -d grid -d leather -d tile -d wood -d bottle -d cable -d capsule -d hazelnut -d metal_nut -d pill -d screw -d toothbrush -d transistor -d zipper mvtec /workspaces/dev/GLASS/dataset/MVTec /workspaces/dev/GLASS/dtd/images

**Comment by ooe1123** [comment]
### mvtec_bottle.onnx

○ main.py
```
@click.option("--num_workers", default=16, type=int, show_default=True)

@main.result_callback()
def run(
        ...
):
    ...
    for dataloader_count, dataloaders in enumerate(list_of_dataloaders):
        ...
```
↓
```
@click.option("--num_workers", default=1, type=int, show_default=True)

@main.result_callback()
def run(
        ...
):
    ...
    for dataloader_count, dataloaders in enumerate(list_of_dataloaders):
        ...
        if dataset_name not in (
            # "mvtec_carpet",
            # "mvtec_grid",
            # "mvtec_leather",
            # "mvtec_tile",
            # "mvtec_wood",
            "mvtec_bottle",
            # "mvtec_cable",
            # "mvtec_capsule",
            # "mvtec_hazelnut",
            # "mvtec_metal_nut",
            # "mvtec_pill",
            # "mvtec_screw",
            # "mvtec_toothbrush",
            # "mvtec_transistor",
            # "mvtec_zipper",
            ):
            continue
```

○ glass.py
```
class GLASS(torch.nn.Module):
    ...
    def set_model_dir(self, model_dir, dataset_name):
        ...
    ...
    def predict(self, test_dataloader):
        ...
        with tqdm.tqdm(test_dataloader, desc="Inferring...", leave=False, unit='batch') as data_iterator:
            for data in data_iterator:
                ...
                if isinstance(data, dict):
                    ...
    ...
    def _predict(self, img):
        ...
        with torch.no_grad():
            ...
            image_scores = self.patch_maker.unpatch_scores(image_scores, batchsize=img.shape[0])
            image_scores = self.patch_maker.score(image_scores)
```
↓
```
flg = False

class GLASS(torch.nn.Module):
    ...
    def set_model_dir(self, model_dir, dataset_name):
        ...
        if dataset_name in (
            # "mvtec_carpet",
            # "mvtec_grid",
            # "mvtec_leather",
            # "mvtec_tile",
            # "mvtec_wood",
            "mvtec_bottle",
            # "mvtec_cable",
            # "mvtec_capsule",
            # "mvtec_hazelnut",
            # "mvtec_metal_nut",
            # "mvtec_pill",
            # "mvtec_screw",
            # "mvtec_toothbrush",
            # "mvtec_transistor",
            # "mvtec_zipper",
            "xxx",
            ):
            global flg
            flg = True
    ...
    def predict(self, test_dataloader):
        ...
        with tqdm.tqdm(test_dataloader, desc="Inferring...", leave=False, unit='batch') as data_iterator:
            for data in data_iterator:
                ...
                if isinstance(data, dict):
                    ...
                if flg:
                    class Exp(torch.nn.Module):
                        def __init__(self, model):
                            super().__init__()
                            self.model = model

                        def forward(self, image):
                            scores, masks = self.model._predict(image)
                            return scores, masks

                    with torch.no_grad():
                        print("------>")
                        model = Exp(self)
                        x = (image,)
                        torch.onnx.export(
                            model, x, 'xxx.onnx',
                            input_names=["images"],
                            output_names=["scores", "masks"],
                            verbose=False, opset_version=17
                        )
                        print("<------")
                        exit()
    ...
    def _predict(self, img):
        ...
        with torch.no_grad():
            ...
            image_scores = self.patch_maker.unpatch_scores(image_scores, batchsize=img.shape[0])
            image_scores = self.patch_maker.score(image_scores)
            if torch.onnx.is_in_onnx_export():
                return image_scores, masks
```

○ common.py
```
class RescaleSegmentor:
    ...
    def convert_to_segmentation(self, patch_scores):
        with torch.no_grad():
            ...
            _scores = _scores.squeeze(1)
```
↓
```
class RescaleSegmentor:
    ...
    def convert_to_segmentation(self, patch_scores):
        with torch.no_grad():
            ...
            _scores = _scores.squeeze(1)
            if torch.onnx.is_in_onnx_export():
                return _scores
```
